### PR TITLE
updates for the "Korg Kaoss DJ" controller script

### PR DESCRIPTION
**Pull request by daschuer**
### New assignments:

|||
|--------------|-----------|
| `browser Knob` | browse library or playlist up & down (same as before) |
| `shift` + `browser Knob` | toggle focus between Playlist and File-Browser |

|||
|--------------|-----------|
| `tap`     |  open folder       |  
| 2 x `tap`     |  close folder       |  
| `(left) shift` + `tap`|  tap bpm of left track | 
| `(right) shift` + `tap`|  tap bpm of right track | 

|||
|--------------|-----------|
| `load A / B` | load track to deck 1 / 2 (same as before) |
| `shift` + `load A / B` | open / close folder in file-browser|   

|||
|--------------|-----------|
| `shift` + `touchpad X` | control super knob of QuickEffectRack for deck 1|   
| `shift` + `touchpad Y` | control super knob of QuickEffectRack for deck 2 |   

### changed assignments
- the loop-on button now works as follows:
   - if a loop is active, it deactivates the loop (same as before)
   - if no loop is active, a new beatloop is created at the current playback position  
    (before a re-loop of the deactivated loop was triggered)

### doc-updates:
https://github.com/mixxxdj/manual/pull/430

### CODE REVIEW

**Comment by daschuer** [comment]
@pi43r Please Test and share your results. Do you have interest to take over?

**Comment by pi43r** [comment]
Thanks for opening this again!
I can try it on my controller but it might take some time. There also seems to be new interest in the [forum](https://mixxx.discourse.group/t/korg-kaoss-dj-midi-mapping-help/16093/41).

I would need to get familiar with scripting for mixxx but I could definitely try updating the docs if I find anything.

**Comment by Swiftb0y** [comment]
@daschuer for me as a reviewer: are you trying to just get this merged or do you want to receive suggestions on potential improvements before merging?

**Comment by daschuer** [comment]
I have no clue about the code here. This PR is just to save the commits from oblivion and that someone else can easily take over.

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:77]: same below
```suggestion
    ([MIDI_CHANNELS.KNOB_LEFT, MIDI_CHANNELS.KNOB_RIGHT])
        .forEach(channel => Object.values(MIDI_DECK_KNOB_LEDS).forEach(
            led => midi.sendShortMsg(channel, led, state)));
```

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:205]:
```suggestion
        const fxGroup = `[EffectRack1_EffectUnit${deck.deckNumber}]`;
```

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:217]:
```suggestion
    if (value === MIDI_ON) {
        if (engine.getValue(deck.group, "loop_enabled");) {
```

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:228]: outdated comment.

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:306]: maybe GoToItem + Back would be better?

**Review by Swiftb0y — COMMENTED** [review]
mergeable otherwise. thank you.

**Comment by Swiftb0y** [comment]
> I have no clue about the code here. This PR is just to save the commits from oblivion and that someone else can easily take over.

Good to know. I doubt anyone else would be interested, so lets fix the easy things and get this merged.

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (daschuer) [reply res/controllers/Korg-KAOSS-DJ-scripts.js:306]: We have no "Back" AFAIK. I leave this for one who can test this.

Comment (Swiftb0y) [reply res/controllers/Korg-KAOSS-DJ-scripts.js:305]: Not really, but we have `[Library], MoveFocusBackward` which is commonly mapped I think.

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:62]: typo.
```suggestion
    this.isFx = false;
```

**Review by Swiftb0y — COMMENTED** [review]
thank you. almost lgtm. Just fix the typo and the fxGroup string above (otherwise it will not get interpreted correctly and `KAOSSDJ.fxTouch` will be broken).

**Comment by daschuer** [comment]
Thank you for review. I have amended the isFX issue. I hope that is OK here.

**Comment by Swiftb0y** [comment]
yup sure. only remaining issue is the unresolved fxGroup issue from my previous review.

**Comment by daschuer** [comment]
Done.

File: res/controllers/Korg-KAOSS-DJ.midi.xml
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ.midi.xml:633]: just guessing here tbh
```suggestion
                <key>KAOSSDJ.fxTouch</key>
                <group>[EffectRack1_EffectUnit1]</group>
```

File: res/controllers/Korg-KAOSS-DJ.midi.xml
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ.midi.xml:643]:
```suggestion
                <key>KAOSSDJ.fxTouch</key>
                <group>[EffectRack1_EffectUnit2]</group>
```

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (Swiftb0y) [inline res/controllers/Korg-KAOSS-DJ-scripts.js:207]:
```suggestion
KAOSSDJ.fxTouch = function(channel, _control, value, _status, group) {
    if (value === MIDI_OFF) {
        engine.setValue(group, "mix", 0);
        engine.setValue(group, "super1", 0);
```

**Review by Swiftb0y — CHANGES_REQUESTED** [review]
good job to eslint for catching this one.

**Review by Swiftb0y — APPROVED** [review]
LGTM. At least this shouldn't have worsened the current version. Lets just merge and see if users notice anything.

**Comment by Swiftb0y** [comment]
@daschuer are you also volunteering to adopt the corresponding manual PR? https://github.com/mixxxdj/manual/pull/430/

**Comment by daschuer** [comment]
No sorry, that needs to be really done by one owning such controller. I think also some testing would be not hurt here.

**Comment by Swiftb0y** [comment]
@raphaelquast do you still have access to the controller and can test the current version of the mapping? Do you have time to finish the manual page?

**Comment by raphaelquast** [comment]
@Swiftb0y 
I do still own the controller but i'm very limited in time at the moment...  (and it seems like it's going to be like this for a while)
If a functionality test from someone owning the controller is the only thing that's blocking this from merging then I can try to find some time to have a quick look... but unfortunately I'm not able to actively contribute in the near future.

**Comment by pi43r** [comment]
I can confirm operation of the new script in Mixxx 4.0.16 on Fedora39.
I couldn't test on MacOs because I'm not able to install the midi driver on my M2 macbook. 
Will edit this for Windows when I get to it.

My findings, most of them pretty minor:
- Moving between levels using Knob [4] + shift needs retriggering of shift after every change 
- Display is off. (Should indicate FX number or param / key or scale. I don't recall this ever to be the case in the past)
- Program Knob [7] does nothing? (supposed to select effect or change key, might be talking about the internal kaoss pad)
- Touchpad [10] does nothing. It should change the super FX knobs. (works with the old script)
- Shift + Play = Key lock not working (doesn't work in old script either)
- Shift + Sync = Cancels Sync does not work. Pressing sync cancels it. (Maybe it's just an issue with the docs?) 


So the main issue is the center touchpad functionality.

**Comment by Swiftb0y** [comment]
Hi @pi43r. Thank you for taking the time to test this mapping.

> Mixxx 4.0.16

Where did you get that version number from, that doesn't look right.

> Moving between levels using Knob [4] + shift needs retriggering of shift after every change

Can you elaborate on what you mean by "retriggering of shift"? What behavior do you expect as working normally? how does it behave instead currently?

> Display is off. (Should indicate FX number or param / key or scale. I don't recall this ever to be the case in the past)

Yeah, I don't think the original mapper bothered to try to understand how that works. Someone with access to the hardware and some basic reverse-engineering skills would have to figure that out.

> Program Knob [7] does nothing? (supposed to select effect or change key, might be talking about the internal kaoss pad)

Thank you for your testing. I'll try to see what I can fix remotely. In the meantime, I see you have already opened https://github.com/mixxxdj/manual/pull/613. Would you be interested in incorporating the changes made here into that PR as well?

**Comment by Swiftb0y** [comment]
> Shift + Play = Key lock not working (doesn't work in old script either)
Shift + Sync = Cancels Sync does not work. Pressing sync cancels it. (Maybe it's just an issue with the docs?)

Those currently seem to be unimplemented for the sake of simplicity. Do you think they're necessary or can we just cut that feature from the docs?

**Review by Swiftb0y — CHANGES_REQUESTED** [review]

**Comment by pi43r** [comment]
> > Mixxx 4.0.16
> Where did you get that version number from, that doesn't look right.
Copied it from human memory :) It should be 2.4.0 (0.16.beta.20240209gitd112e4c.fc39) 
https://koji.rpmfusion.org/koji/buildinfo?buildID=28064

> > Moving between levels using Knob [4] + shift needs retriggering of shift after every change
> 
> Can you elaborate on what you mean by "retriggering of shift"? What behavior do you expect as working normally? how does it behave instead currently?
Holding shift + knob 4 changes from Browser to Search to the Folder window. This change only happens once. I have to release shift and press it again in order to highlight the next "level".  I would expect it to cycle through them as long as it shift is pressed.


> Would you be interested in incorporating the changes made here into that PR as well?
Sure, I can edit the manual

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (pi43r) [reply res/controllers/Korg-KAOSS-DJ-scripts.js:184]: confirmed working 👍

File: res/controllers/Korg-KAOSS-DJ-scripts.js
Comment (pi43r) [reply res/controllers/Korg-KAOSS-DJ-scripts.js:189]: Just checked. Yes it is
Status: 0xB6 Control: 0x1F Options: 0x200

Comment (daschuer) [reply res/controllers/Korg-KAOSS-DJ-scripts.js:189]: Yes, then you may issue a PR to my daschuer/mixxx repository and I can integrated here. 

Or create your own pull request against mixxxdj/mixxx and we can close this PR

**Review by Swiftb0y — APPROVED** [review]
LGTM. Thank you.

**Comment by Swiftb0y** [comment]
I partly wrote this PR @daschuer so I'll let you decide whether to merge.

**Comment by daschuer** [comment]
Thank you. I see no reason to hold that back.

**Comment by raphaelquast** [comment]
@all really nice to see that the changes for to the Korg Kaoss DJ script have made it to the finish line! Thanks for following up on my initial PR (and sorry for not being able to finish it myself in reasonable time)!

**Comment by daschuer** [comment]
@pi43r I have noticed we have not ask you for the formal confirmation to distribute your work  with Mixxx. 
Can you please sign https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ and comment here when done? 
Thank you.

**Comment by ronso0** [comment]
@pi43r please sign the CLA (see link above) and let us know when done.
Thank you!

**Comment by pi43r** [comment]
sorry, I thought I already signed it. Did it again just now!

**Pull request merged**
